File: app/src/main/java/com/github/shadowsocks/ShadowsocksApplication.java
package com.github.shadowsocks;


import android.annotation.SuppressLint;
import android.app.Application;
import android.content.SharedPreferences;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;
import android.preference.PreferenceManager;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;

import com.evernote.android.job.JobManager;
import com.github.shadowsocks.database.DBHelper;
import com.github.shadowsocks.database.Profile;
import com.github.shadowsocks.database.ProfileManager;
import com.github.shadowsocks.database.SSRSubManager;
import com.github.shadowsocks.job.DonaldTrump;
import com.github.shadowsocks.utils.Constants;
import com.github.shadowsocks.utils.IOUtils;
import com.github.shadowsocks.utils.TcpFastOpen;
import com.github.shadowsocks.utils.ToastUtils;
import com.github.shadowsocks.utils.Utils;
import com.github.shadowsocks.utils.VayLog;
import com.google.android.gms.analytics.GoogleAnalytics;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.StandardExceptionParser;
import com.google.android.gms.analytics.Tracker;
import com.google.android.gms.common.api.PendingResult;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.tagmanager.Container;
import com.google.android.gms.tagmanager.ContainerHolder;
import com.google.android.gms.tagmanager.TagManager;
import com.j256.ormlite.logger.LocalLog;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;

import eu.chainfire.libsuperuser.Shell;

public class ShadowsocksApplication extends Application {
    public static ShadowsocksApplication app;

    private static final String TAG = ShadowsocksApplication.class.getSimpleName();
    public static final String SIG_FUNC = "getSignature";

    private String[] EXECUTABLES = {
            Constants.Executable.PDNSD,
            Constants.Executable.REDSOCKS,
            Constants.Executable.SS_TUNNEL,
            Constants.Executable.SS_LOCAL,
            Constants.Executable.TUN2SOCKS,
            Constants.Executable.KCPTUN};

    /**
     * The ones in Locale doesn't have script included
     */
    private static final Locale SIMPLIFIED_CHINESE;
    private static final Locale TRADITIONAL_CHINESE;

    static {
        if (Build.VERSION.SDK_INT >= 21) {
            SIMPLIFIED_CHINESE = Locale.forLanguageTag("zh-Hans-CN");
            TRADITIONAL_CHINESE = Locale.forLanguageTag("zh-Hant-TW");
        } else {
            SIMPLIFIED_CHINESE = Locale.SIMPLIFIED_CHINESE;
            TRADITIONAL_CHINESE = Locale.TRADITIONAL_CHINESE;
        }
    }

    public ContainerHolder containerHolder;

    private Tracker tracker;
    public SharedPreferences settings;
    public SharedPreferences.Editor editor;

    public ProfileManager profileManager;
    public SSRSubManager ssrsubManager;
    public Resources resources;

    public boolean isNatEnabled() {
        return false;
    }


    public boolean isVpnEnabled() {
        return !isNatEnabled();
    }

    public ScheduledExecutorService mThreadPool;

    /**
     * /// xhao: init variable
     */
    private void initVariable() {
        tracker = GoogleAnalytics.getInstance(this).newTracker(R.xml.tracker);
        settings = PreferenceManager.getDefaultSharedPreferences(this);
        editor = settings.edit();

        profileManager = new ProfileManager(new DBHelper(this));
        ssrsubManager = new SSRSubManager(new DBHelper(this));
        resources = getResources();

        mThreadPool = new ScheduledThreadPoolExecutor(10, new ThreadFactory() {
            @Override
            public Thread newThread(@NonNull Runnable r) {
                Thread thread = new Thread(r);
                thread.setName("shadowsocksr-thread");
                return thread;
            }
        });
    }

    /**
     * send event
     */
    public void track(String category, String action) {
        Map<String, String> builders = new HitBuilders.EventBuilder()
                .setAction(action)
                .setCategory(category)
                .setLabel(BuildConfig.VERSION_NAME)
                .build();
        tracker.send(builders);
    }

    /**
     * send event
     */
    public void track(Throwable t) {
        Map<String, String> builders = new HitBuilders.ExceptionBuilder()
                .setDescription(new StandardExceptionParser(this, null).getDescription(Thread.currentThread().getName(), t))
                .setFatal(false)
                .build();
        tracker.send(builders);
    }

    /**
     * get profile id
     *
     * @return no save return -1
     */
    public int profileId() {
        return settings.getInt(Constants.Key.id, -1);
    }

    /**
     * save profile id
     *
     * @param i profile id
     */
    public void profileId(int i) {
        editor.putInt(Constants.Key.id, i).apply();
    }

    /**
     * current profile
     */
    public Profile currentProfile() {
        return profileManager.getProfile(profileId());
    }

    /**
     * switch profile
     *
     * @param id profile id
     */
    public Profile switchProfile(int id) {
        profileId(id);

        Profile profile = profileManager.getProfile(id);
        if (profile != null) {
            return profile;
        } else {
            return profileManager.createProfile();
        }
    }

    @SuppressLint("NewApi")
    private Locale checkChineseLocale(Locale locale) {
        if ("zh".equals(locale.getLanguage())) {
            String country = locale.getCountry();
            if ("CN".equals(country) || "TW".equals(country)) {
                return null;
            } else {
                String script = locale.getScript();
                if ("Hans".equals(script)) {
                    return SIMPLIFIED_CHINESE;
                } else if ("Hant".equals(script)) {
                    return TRADITIONAL_CHINESE;
                } else {
                    VayLog.w(TAG, String.format("Unknown zh locale script: %s. Falling back to trying countries...", script));
                    if ("SG".equals(country)) {
                        return SIMPLIFIED_CHINESE;
                    } else if ("HK".equals(country) || "MO".equals(country)) {
                        return TRADITIONAL_CHINESE;
                    } else {
                        VayLog.w(TAG, String.format("Unknown zh locale: %s. Falling back to zh-Hans-CN...", locale.toLanguageTag()));
                        return SIMPLIFIED_CHINESE;
                    }
                }
            }
        } else {
            return null;
        }
    }

    /**
     * check chinese locale
     */
    private void checkChineseLocale(Configuration config) {
        if (Build.VERSION.SDK_INT >= 24) {
            LocaleList localeList = config.getLocales();
            Locale[] newList = new Locale[localeList.size()];
            boolean changed = false;
            for (int i = 0; i < localeList.size(); i++) {
                Locale locale = localeList.get(i);
                Locale newLocale = checkChineseLocale(locale);
                if (newLocale == null) {
                    newList[i] = locale;
                } else {
                    newList[i] = newLocale;
                    changed = true;
                }
            }
            if (changed) {
                Configuration newConfig = new Configuration(config);
                newConfig.setLocales(new LocaleList(newList));
                Resources res = getResources();
                res.updateConfiguration(newConfig, res.getDisplayMetrics());
            }
        } else {
            Locale newLocale = checkChineseLocale(config.locale);
            if (newLocale != null) {
                Configuration newConfig = new Configuration(config);
                newConfig.locale = newLocale;
                Resources res = getResources();
                res.updateConfiguration(newConfig, res.getDisplayMetrics());
            }
        }
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        checkChineseLocale(newConfig);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        app = this;
        // init toast utils
        ToastUtils.init(getApplicationContext());
        initVariable();

        if (!BuildConfig.DEBUG) {
            java.lang.System.setProperty(LocalLog.LOCAL_LOG_LEVEL_PROPERTY, "ERROR");
        }
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
        checkChineseLocale(getResources().getConfiguration());
        TagManager tm = TagManager.getInstance(this);
        PendingResult<ContainerHolder> pending = tm.loadContainerPreferNonDefault("GTM-NT8WS8", R.raw.gtm_default_container);
        ResultCallback<ContainerHolder> callback = new ResultCallback<ContainerHolder>() {
            @Override
            public void onResult(@NonNull ContainerHolder holder) {
                if (!holder.getStatus().isSuccess()) {
                    return;
                }
                containerHolder = holder;
                Container container = holder.getContainer();
                container.registerFunctionCallMacroCallback(SIG_FUNC, new Container.FunctionCallMacroCallback() {
                    @Override
                    public Object getValue(String functionName, Map<String, Object> parameters) {
                        if (SIG_FUNC.equals(functionName)) {
                            return Utils.getSignature(getApplicationContext());
                        } else {
                            return null;
                        }
                    }
                });
            }
        };
        pending.setResultCallback(callback, 2, TimeUnit.SECONDS);
        JobManager.create(this).addJobCreator(new DonaldTrump());

        if (settings.getBoolean(Constants.Key.tfo, false) && TcpFastOpen.supported()) {
            mThreadPool.execute(new Runnable() {
                @Override
                public void run() {
                    TcpFastOpen.enabled(settings.getBoolean(Constants.Key.tfo, false));
                }
            });
        }
    }

    /**
     * refresh container holder
     */
    public void refreshContainerHolder() {
        if (containerHolder != null) {
            containerHolder.refresh();
        }
    }

    /**
     * copy assets
     *
     * @param path assets path
     */
    private void copyAssets(String path) {
        AssetManager assetManager = getAssets();
        String[] files = null;
        try {
            files = assetManager.list(path);
        } catch (Exception e) {
            VayLog.e(TAG, e.getMessage());
            app.track(e);
        }

        if (files != null) {
            for (String file : files) {
                InputStream in = null;
                FileOutputStream fos = null;
                try {
                    if (!TextUtils.isEmpty(path)) {
                        in = assetManager.open(path + File.separator + file);
                    } else {
                        in = assetManager.open(file);
                    }
                    fos = new FileOutputStream(getApplicationInfo().dataDir + '/' + file);
                    IOUtils.copy(in, fos);
                } catch (IOException e) {
                    VayLog.e(TAG, "copyAssets", e);
                } finally {
                    try {
                        if (in != null) {
                            in.close();
                        }
                    } catch (Exception e) {
                        VayLog.e(TAG, "copyAssets", e);
                    }

                    try {
                        if (fos != null) {
                            fos.close();
                        }
                    } catch (Exception e) {
                        VayLog.e(TAG, "copyAssets", e);
                    }
                }
            }
        }
    }

    /**
     * arash recovery
     */
    public void crashRecovery() {
        ArrayList<String> cmd = new ArrayList<>();

        String[] paramsArray = {"libssr-local.so", "ss-tunnel", "libpdnsd.so", "libredsocks.so", "libtun2socks.so", "proxychains"};
        for (String task : paramsArray) {
            cmd.add(String.format(Locale.ENGLISH, "killall %s", task));
            cmd.add(String.format(Locale.ENGLISH, "rm -f %1$s/%2$s-nat.conf %1$s/%2$s-vpn.conf", getApplicationInfo().dataDir, task));
        }

        // convert to cmd array
        String[] cmds = convertListToStringArray(cmd);
        if (app.isNatEnabled()) {
            cmd.add("iptables -t nat -F OUTPUT");
            cmd.add("echo done");
            List<String> result = Shell.SU.run(cmds);
            if (result != null && !result.isEmpty()) {
                // fallback to SH
                return;
            }
        }

        Shell.SH.run(cmds);
    }

    /**
     * convert list to string array
     *
     * @param list list
     * @return convert failed return {}
     */
    private String[] convertListToStringArray(List<String> list) {
        if (list == null || list.isEmpty()) {
            return new String[]{};
        }

        // start convert
        String[] result = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            result[i] = list.get(i);
        }
        return result;
    }

    /**
     * copy assets
     */
    public void copyAssets() {
        // ensure executables are killed before writing to them
        crashRecovery();
        copyAssets(System.getABI());
        copyAssets("acl");

        // exec cmds
        String[] cmds = new String[EXECUTABLES.length];
        for (int i = 0; i < cmds.length; i++) {
            cmds[i] = "chmod 755 " + getApplicationInfo().nativeLibraryDir + File.separator + EXECUTABLES[i];
        }
        Shell.SH.run(cmds);

        // save current version code
        editor.putInt(Constants.Key.currentVersionCode, BuildConfig.VERSION_CODE).apply();
    }

    /**
     * update assets
     */
    public void updateAssets() {
        if (settings.getInt(Constants.Key.currentVersionCode, -1) != BuildConfig.VERSION_CODE) {
            copyAssets();
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/ShadowsocksApplication.kt
package com.bige0.shadowsocksr

import android.annotation.*
import android.app.*
import android.content.*
import android.content.res.*
import android.os.*
import android.preference.*
import androidx.appcompat.app.*
import com.bige0.shadowsocksr.database.*
import com.bige0.shadowsocksr.job.*
import com.bige0.shadowsocksr.utils.*
import com.evernote.android.job.*
import com.google.android.gms.analytics.*
import com.google.android.gms.common.api.*
import com.google.android.gms.tagmanager.*
import com.j256.ormlite.logger.*
import java.io.*
import java.util.*
import java.util.concurrent.*

class ShadowsocksApplication : Application()
{
	companion object
	{
		const val SIG_FUNC = "getSignature"
		private const val TAG = "ShadowsocksApplication"
		private val EXECUTABLES = arrayOf(Constants.Executable.PDNSD, Constants.Executable.PROXYCHAINS4, Constants.Executable.SS_LOCAL, Constants.Executable.TUN2SOCKS)

		private val SIMPLIFIED_CHINESE: Locale
		private val TRADITIONAL_CHINESE: Locale
		lateinit var app: ShadowsocksApplication

		init
		{
			if (Build.VERSION.SDK_INT >= 21)
			{
				SIMPLIFIED_CHINESE = Locale.forLanguageTag("zh-Hans-CN")
				TRADITIONAL_CHINESE = Locale.forLanguageTag("zh-Hant-TW")
			}
			else
			{
				SIMPLIFIED_CHINESE = Locale.SIMPLIFIED_CHINESE
				TRADITIONAL_CHINESE = Locale.TRADITIONAL_CHINESE
			}
		}
	}

	lateinit var containerHolder: ContainerHolder
	lateinit var settings: SharedPreferences
	lateinit var editor: SharedPreferences.Editor
	lateinit var profileManager: ProfileManager
	lateinit var ssrSubManager: SSRSubManager
	lateinit var mThreadPool: ScheduledExecutorService
	private lateinit var tracker: Tracker

	private fun initVariable()
	{
		tracker = GoogleAnalytics.getInstance(this)
			.newTracker(R.xml.tracker)
		settings = PreferenceManager.getDefaultSharedPreferences(this)
		editor = settings.edit()

		profileManager = ProfileManager(DBHelper(this))
		ssrSubManager = SSRSubManager(DBHelper(this))

		mThreadPool = ScheduledThreadPoolExecutor(10, ThreadFactory { r ->
			val thread = Thread(r)
			thread.name = "shadowsocksr-thread"
			thread
		})
	}

	fun track(category: String, action: String)
	{
		val builders = HitBuilders.EventBuilder()
			.setAction(action)
			.setCategory(category)
			.setLabel(BuildConfig.VERSION_NAME)
			.build()
		tracker.send(builders)
	}

	fun track(t: Throwable)
	{
		val builders = HitBuilders.ExceptionBuilder()
			.setDescription(StandardExceptionParser(this, null).getDescription(Thread.currentThread().name, t))
			.setFatal(false)
			.build()
		tracker.send(builders)
	}

	fun profileId(): Int
	{
		return settings.getInt(Constants.Key.id, -1)
	}

	fun profileId(i: Int)
	{
		editor.putInt(Constants.Key.id, i)
			.apply()
	}

	fun currentProfile(): Profile?
	{
		return profileManager.getProfile(profileId())
	}

	fun switchProfile(id: Int): Profile
	{
		profileId(id)
		val profile = profileManager.getProfile(id)
		return profile ?: profileManager.createProfile()
	}

	@SuppressLint("NewApi")
	private fun checkChineseLocale(locale: Locale): Locale?
	{
		if ("zh" == locale.language)
		{
			val country = locale.country
			if ("CN" == country || "TW" == country)
			{
				return null
			}
			val script = locale.script
			if ("Hans" == script)
			{
				return SIMPLIFIED_CHINESE
			}
			else if ("Hant" == script)
			{
				return TRADITIONAL_CHINESE
			}
			VayLog.w(TAG, String.format("Unknown zh locale script: %s. Falling back to trying countries...", script))
			return if ("SG" == country)
			{
				SIMPLIFIED_CHINESE
			}
			else if ("HK" == country || "MO" == country)
			{
				TRADITIONAL_CHINESE
			}
			else
			{
				VayLog.w(TAG, String.format("Unknown zh locale: %s. Falling back to zh-Hans-CN...", locale.toLanguageTag()))
				SIMPLIFIED_CHINESE
			}
		}
		return null
	}

	private fun checkChineseLocale(config: Configuration)
	{
		if (Build.VERSION.SDK_INT >= 24)
		{
			val localeList = config.locales
			val newList = arrayOfNulls<Locale>(localeList.size())
			var changed = false
			for (i in 0 until localeList.size())
			{
				val locale = localeList.get(i)
				val newLocale = checkChineseLocale(locale)
				if (newLocale == null)
				{
					newList[i] = locale
				}
				else
				{
					newList[i] = newLocale
					changed = true
				}
			}
			if (changed)
			{
				val newConfig = Configuration(config)
				newConfig.setLocales(LocaleList(*newList))
				resources.updateConfiguration(newConfig, resources.displayMetrics)
			}
		}
		else
		{
			val newLocale = checkChineseLocale(config.locale)
			if (newLocale != null)
			{
				val newConfig = Configuration(config)
				newConfig.locale = newLocale
				resources.updateConfiguration(newConfig, resources.displayMetrics)
			}
		}
	}

	override fun onConfigurationChanged(newConfig: Configuration)
	{
		super.onConfigurationChanged(newConfig)
		checkChineseLocale(newConfig)
	}

	override fun onCreate()
	{
		super.onCreate()
		app = this
		initVariable()

		if (!BuildConfig.DEBUG)
		{
			System.setProperty(LocalLog.LOCAL_LOG_LEVEL_PROPERTY, "ERROR")
		}
		AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
		checkChineseLocale(resources.configuration)
		val tm = TagManager.getInstance(this)
		val pending = tm.loadContainerPreferNonDefault("GTM-NT8WS8", R.raw.gtm_default_container)
		val callback = ResultCallback<ContainerHolder> { holder ->
			if (!holder.status.isSuccess)
			{
				return@ResultCallback
			}
			containerHolder = holder
			val container = holder.container
			container.registerFunctionCallMacroCallback(SIG_FUNC) { functionName, _ ->
				if (SIG_FUNC == functionName)
				{
					Utils.getSignature(applicationContext)
				}
				else
				{
					null
				}
			}
		}
		pending.setResultCallback(callback, 2, TimeUnit.SECONDS)
		JobManager.create(this)
			.addJobCreator(DonaldTrump())
	}

	fun refreshContainerHolder()
	{
		if (::containerHolder.isInitialized)
		{
			containerHolder.refresh()
		}
	}

	private fun copyAssets(path: String)
	{
		val assetManager = assets
		var files: Array<String>? = null
		try
		{
			files = assetManager.list(path)
		}
		catch (e: Exception)
		{
			VayLog.e(TAG, e.message.toString())
			app.track(e)
		}

		if (files != null)
		{
			for (file in files)
			{
				var inputStream: InputStream? = null
				var fos: FileOutputStream? = null
				try
				{
					inputStream = if (path.isNotEmpty())
					{
						assetManager.open(path + File.separator + file)
					}
					else
					{
						assetManager.open(file)
					}
					fos = FileOutputStream("${applicationInfo.dataDir}/$file")
					IOUtils.copy(inputStream, fos)
				}
				catch (e: IOException)
				{
					VayLog.e(TAG, "copyAssets", e)
				}
				finally
				{
					try
					{
						inputStream?.close()
					}
					catch (e: Exception)
					{
						VayLog.e(TAG, "copyAssets", e)
					}
					try
					{
						fos?.close()
					}
					catch (e: Exception)
					{
						VayLog.e(TAG, "copyAssets", e)
					}
				}
			}
		}
	}

	fun crashRecovery()
	{
		for (exe in EXECUTABLES)
		{
			ProcessBuilder("sh", "killall", exe).start()
			ProcessBuilder("sh", "rm", "-f", "${applicationInfo.dataDir}${File.separator}${exe}-vpn.conf").start()
		}
	}

	fun copyAssets()
	{
		// ensure executables are killed before writing to them
		crashRecovery()
		copyAssets(Build.SUPPORTED_ABIS[0])
		copyAssets("acl")

		for (exe in EXECUTABLES)
		{
			ProcessBuilder("sh", "chmod", "755", "${applicationInfo.nativeLibraryDir}${File.separator}${exe}").start()
		}

		// save current version code
		editor.putInt(Constants.Key.currentVersionCode, BuildConfig.VERSION_CODE)
			.apply()
	}

	fun updateAssets()
	{
		if (settings.getInt(Constants.Key.currentVersionCode, -1) != BuildConfig.VERSION_CODE)
		{
			copyAssets()
		}
	}
}
